### PR TITLE
Fetch block number when forking latest block

### PR DESCRIPTION
**Pull request by war-in**
Partially solves #795 

## Introduced changes

- `BlockId::Tag(Latest)` is converted to the `BlockId::Number(...)` in running.rs

## Checklist

- [x] Linked relevant issue
- [ ] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [ ] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: crates/cheatnet/src/forking/state.rs [hunk 1]
```diff
@@ -200,3 +211,10 @@ impl StateReader for ForkStateReader {
         ))
     }
 }
+
+fn get_latest_block_number(client: &JsonRpcClient<HttpTransport>, runtime: &Runtime) -> BlockId {
+    return match runtime.block_on(client.get_block_with_tx_hashes(BlockId::Tag(Latest))) {
+        Ok(MaybePendingBlockWithTxHashes::Block(block)) => BlockId::Number(block.block_number),
+        _ => panic!("Could not get the latest block number"),
```
[the remark this answers is not in the document]
Comment (war-in) [reply crates/cheatnet/src/forking/state.rs:218]: network error/ rpc error (basically anything)

Ok, but when should we unwrap it?

[the remark this answers is not in the document]
Comment (piotmag769) [reply crates/cheatnet/src/forking/state.rs:218]: Never, it should be bubbled using `?` up to the runner

**Review by piotmag769 — APPROVED** [review]

File: crates/cheatnet/src/forking/state.rs [hunk 2]
```diff
@@ -200,3 +211,13 @@ impl StateReader for ForkStateReader {
         ))
     }
 }
+
+fn get_latest_block_number(
+    client: &JsonRpcClient<HttpTransport>,
+    runtime: &Runtime,
+) -> Result<BlockId> {
+    return match runtime.block_on(client.get_block_with_tx_hashes(BlockId::Tag(Latest))) {
```
Comment (Arcticae) [inline crates/cheatnet/src/forking/state.rs:219]: nit: you don't need the `return`

**Review by Arcticae — COMMENTED** [review]
Wouldn't it be better, if we instead passed a block number which is pre-fetched to the `ForkStateReader`, and calculate/fetch the value outside the class? I think it'd make for a cleaner interface and code

**Comment by piotmag769** [comment]
@Arcticae I think we want to avoid spawning random runtimes in forge code

**Comment by Arcticae** [comment]
Also, this will save the cache in the `Latest` file, instead of a block number, we should probably fix that

File: crates/forge/src/running.rs [hunk 3]
```diff
@@ -229,48 +234,71 @@ fn get_fork_state_reader(
...
-) -> Option<ForkStateReader> {
+) -> Result<Option<ForkStateReader>> {
     match &fork_config {
-        Some(ForkConfig::Params(url, block_id)) => Some(ForkStateReader::new(
-            url,
-            *block_id,
-            Some(workspace_root.join(".snfoundry_cache").as_ref()),
-        )),
+        Some(ForkConfig::Params(url, block_id)) => {
+            let block_id = get_block_id(url, *block_id)?;
```
Comment (Arcticae) [inline crates/forge/src/running.rs:240]: I would just name the function `get_latest_block id` and only invoke it `if let BlockIdTag(Latest) = block_id`

**Review by Arcticae — APPROVED** [review]
Much better

**Pull request merged**